rust/arrow/src/compute/cast_kernels.rs

```diff
@@ -0,0 +1,336 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines cast kernels for `ArrayRef`.
+//!
+//! Allows casting arrays between supported datatypes.
+
+use std::sync::Arc;
+
+use crate::array::*;
+use crate::builder::*;
+use crate::datatypes::*;
+use crate::error::{ArrowError, Result};
+
+// TODO:
+//  * remove downcast unwraps and replace with explicit errors
+
+/// Macro rule to cast between numeric types
+macro_rules! cast_numeric_arrays {
+    ($array:expr, $from_ty:ident, $to_ty:ident) => {{
+        match numeric_cast::<$from_ty, $to_ty>(
+            $array
+                .as_any()
+                .downcast_ref::<PrimitiveArray<$from_ty>>()
+                .unwrap(),
+        ) {
+            Ok(to) => Ok(Arc::new(to) as ArrayRef),
+            Err(e) => Err(e),
+        }
+    }};
+}
+
+macro_rules! cast_numeric_to_string {
+    ($array:expr, $from_ty:ident) => {{
+        match cast_numeric_to_string::<$from_ty>(
+            $array
+                .as_any()
+                .downcast_ref::<PrimitiveArray<$from_ty>>()
+                .unwrap(),
+        ) {
+            Ok(to) => Ok(Arc::new(to) as ArrayRef),
+            Err(e) => Err(e),
+        }
+    }};
+}
+
+macro_rules! cast_string_to_numeric {
+    ($array:expr, $to_ty:ident) => {{
+        match cast_string_to_numeric::<$to_ty>(
+            $array.as_any().downcast_ref::<BinaryArray>().unwrap(),
+        ) {
+            Ok(to) => Ok(Arc::new(to) as ArrayRef),
+            Err(e) => Err(e),
+        }
+    }};
+}
+
+/// Cast array to provided data type
+pub fn cast(array: &ArrayRef, to_type: &DataType) -> Result<ArrayRef> {
+    use DataType::*;
+    let from_type = array.data_type();
+
+    // clone array if types are the same
+    if from_type == to_type {
+        return Ok(array.clone());
+    }
+    match (from_type, to_type) {
+        (Struct(_), _) => Err(ArrowError::ComputeError(
+            "Cannot cast from struct to other types".to_string(),
+        )),
+        (_, Struct(_)) => Err(ArrowError::ComputeError(
+            "Cannot cast to struct from other types".to_string(),
+        )),
+        (List(_), List(_)) => unimplemented!("Casting between lists not yet supported"),
+        (List(_), _) => Err(ArrowError::ComputeError(
+            "Cannot cast list to non-list data types".to_string(),
+        )),
+        (_, List(_)) => unimplemented!("Casting scalars to lists not yet supported"),
```

Comment:
It should be easy to change these `unimplemented!` calls into `Err` now?

Reply:
All the places where I'm not already returning an error are placeholders. I'll replace the `unimplemented` with cast functions